### ISSUE TITLE
some problems about experiments

### ISSUE THREAD

**Issue by freeman-1995** [lorenmt/shape-adaptor#2]
Hi

I have some problems about some results in experiments.

1. I just wonder if you remove one branch of shape adaptor, the trained network will maintain the same performance.

2. In imagenet experiment, the search space of shape adaptor is 0.5~1.0, it is larger than original setting in resnet50, so the performance from 77.18±0.04  to 78.74±0.12 is acquired by larger search space.

3. I just curious, if I remove one branch of shape adaptor and retrain the searched architecture,  can I get the same performance as previous?

4. I am interested in AUTOSC, in mobilenetv2 experiments, autosc's result is larger both in Params and MACS, so the performance is better, can autosc outperform original setting when it has smaller both in Params and MACS?

I try to run model_training.py in distributed data parallel training without changing any hyperparameters. I trained the vgg16 with 8 P100 GPUS and test on valid dataset. The test acc is not so good compared to the vgg16 trained from scratch. I just wonder whether the result of experiments in paper are with retrain?

2020-09-13 07:08:44,075:INFO:Epoch:199 step:1200 lr:6.168375916970619e-06 loss:2.165580 acc:0.520218 fps:178.7379645282133
2020-09-13 07:08:49,295:INFO:Epoch:199 step:1210 lr:6.168375916970619e-06 loss:2.165537 acc:0.520227 fps:196.1497854095874
2020-09-13 07:08:55,025:INFO:Epoch:199 step:1220 lr:6.168375916970619e-06 loss:2.165494 acc:0.520237 fps:178.72407086182486
2020-09-13 07:09:00,241:INFO:Epoch:199 step:1230 lr:6.168375916970619e-06 loss:2.165451 acc:0.520246 fps:196.33716160952534
2020-09-13 07:09:06,005:INFO:Epoch:199 step:1240 lr:6.168375916970619e-06 loss:2.165404 acc:0.520256 fps:177.68041829805352
2020-09-13 07:09:11,207:INFO:Epoch:199 step:1250 lr:6.168375916970619e-06 loss:2.165362 acc:0.520264 fps:196.83474920999691

EPOCH: 0199 ITER: 250400 | TRAIN [LOSS|ACC.]: 1.1114 0.7423 || TEST [LOSS|ACC.]: 1.8600 0.5766 || MACs 7.477G Params 15.236M
s(alpha) = [array([0.99647926, 0.9964938 , 0.93286162, 0.55869093])] | current shape = [32, 31, 31, 31, 30, 30, 30, 27, 27, 27, 15, 15, 15]
TOP: 0.5766369700431824
Total training takes 137928.8354 seconds.
EPOCH: 0199 ITER: 250400 | TRAIN [LOSS|ACC.]: 1.1098 0.7434 || TEST [LOSS|ACC.]: 1.8378 0.5805 || MACs 7.477G Params 15.236M
s(alpha) = [array([0.99647926, 0.9964938 , 0.93286162, 0.55869093])] | current shape = [32, 31, 31, 31, 30, 30, 30, 27, 27, 27, 15, 15, 15]
TOP: 0.5811771154403687
Total training takes 137928.9651 seconds.
EPOCH: 0199 ITER: 250400 | TRAIN [LOSS|ACC.]: 1.1087 0.7443 || TEST [LOSS|ACC.]: 1.8509 0.5730 || MACs 7.477G Params 15.236M
s(alpha) = [array([0.99647926, 0.9964938 , 0.93286162, 0.55869093])] | current shape = [32, 31, 31, 31, 30, 30, 30, 27, 27, 27, 15, 15, 15]
TOP: 0.5729699730873108
Total training takes 137929.0455 seconds.
EPOCH: 0199 ITER: 250400 | TRAIN [LOSS|ACC.]: 1.1104 0.7426 || TEST [LOSS|ACC.]: 1.8716 0.5697 || MACs 7.477G Params 15.236M
s(alpha) = [array([0.99647926, 0.9964938 , 0.93286162, 0.55869093])] | current shape = [32, 31, 31, 31, 30, 30, 30, 27, 27, 27, 15, 15, 15]
TOP: 0.5728103518486023
Total training takes 137929.0534 seconds.
EPOCH: 0199 ITER: 250400 | TRAIN [LOSS|ACC.]: 1.1038 0.7453 || TEST [LOSS|ACC.]: 1.8506 0.5799 || MACs 7.477G Params 15.236M
s(alpha) = [array([0.99647926, 0.9964938 , 0.93286162, 0.55869093])] | current shape = [32, 31, 31, 31, 30, 30, 30, 27, 27, 27, 15, 15, 15]
TOP: 0.5799017548561096
Total training takes 137929.0631 seconds.
EPOCH: 0199 ITER: 250400 | TRAIN [LOSS|ACC.]: 1.1137 0.7435 || TEST [LOSS|ACC.]: 1.9037 0.5683 || MACs 7.477G Params 15.236M
s(alpha) = [array([0.99647926, 0.9964938 , 0.93286162, 0.55869093])] | current shape = [32, 31, 31, 31, 30, 30, 30, 27, 27, 27, 15, 15, 15]
TOP: 0.569606363773346
Total training takes 137929.1492 seconds.
EPOCH: 0199 ITER: 250400 | TRAIN [LOSS|ACC.]: 1.1027 0.7443 || TEST [LOSS|ACC.]: 1.8845 0.5733 || MACs 7.477G Params 15.236M
s(alpha) = [array([0.99647926, 0.9964938 , 0.93286162, 0.55869093])] | current shape = [32, 31, 31, 31, 30, 30, 30, 27, 27, 27, 15, 15, 15]
TOP: 0.5741389989852905
Total training takes 137929.1914 seconds.
EPOCH: 0199 ITER: 250400 | TRAIN [LOSS|ACC.]: 1.1115 0.7428 || TEST [LOSS|ACC.]: 1.8278 0.5795 || MACs 7.477G Params 15.236M
s(alpha) = [array([0.99647926, 0.9964938 , 0.93286162, 0.55869093])] | current shape = [32, 31, 31, 31, 30, 30, 30, 27, 27, 27, 15, 15, 15]

**Comment by lorenmt** [comment]
Could you further elaborate which dataset did you run? Both CIFAR10/100 should not give such low accuracy. Considering your training accuracy is even not in 100%, I highly doubt there exists some bugs in your code. 

I would suggest to run my clean single GPU version first, without any modification, to at least have a general intuition how this model would behave. It would be really helpful to further modify it into your own research correctly.

For the default CIFAR-100, the model should converge within 24 hours in P100.

I will answer your other questions tomorrow.

**Comment by freeman-1995** [comment]
@lorenmt hi~
I run it on ILSVRC2012_imagnet dataset with 8 P100 becaues single GPU is too slow, I  modeifid it to parallel training. I'll run it again on cifar10 with single GPU

**Comment by lorenmt** [comment]
Ok. For ImagetNet, please remember to change input_dim flag into 224. You can also check the print log: current shape, to make sure the model works as you expected.

The default hyper-parameters in this repo is for small resolution dataset (explained further in readme). You may check my hyper-parameter setting in all datasets from the appendix of the paper.

Anyway, this incorrect 32 resolution imagenet still indicates the shape adaptor gives 58% accuracy higher than 52% from original network. So I still not quite follow your concern here?

**Comment by freeman-1995** [comment]
oh, resolution is not correct, I made a mistake, I appreciate your nice reply!! I check and run it again now,  and the questitions above  I hope you can give some thoughts when you are free. thanks

**Comment by lorenmt** [comment]
For your questions:

1. If you re-train the network with a fixed shape, the performance would drop a little bit, but still much better than the human-designed network. This might be due to this adaptive training helps learn spatial-invariant features.

2. Yes, you can say the improved performance is due to a larger shape. However, a larger shape would not guarantee an improved performance: A very large shape in some cases can perform even worse than the human-designed shape, though the memory is a lot higher. Please see Section 4.4 in the paper for more analysis.

3. I am not sure what's the difference between this question from the first one?

4. In AutoSC, we changed the width of the network in order to (roughly) match the memory cost of the shape adaptor network in each dataset. In the implementation, since we only search for the shape of the network, if in order to match both parameter space and memory space of the human-designed network, this network could only be the human-designed network itself.

**Comment by freeman-1995** [comment]
ok, good,  thanks for your nice answer

**Comment by freeman-1995** [comment]
@lorenmt hi~ 
I still have some questions about section4.4. I find that the model generated from standard needs no retrain, this is surprising and very valuable in NAS. The reason explained in the paper is thanks to the dynamically updating network, which helps to get spatial-invariant features, I can't get it, could you explain more in details about spatial-invariant features?
section4.4, all experiments is on cifar100 and aircraft, how about imagenet's result? can standard training still need no retrain?

**Comment by lorenmt** [comment]
Yes, all experiments were done without re-training, and the conclusion is consistent across all networks and datasets.

Spatial-invariant features mean the features are invariant to different spatial size of input data. Since during training, the network shape is changing, the features are then updated based on changing network shapes, i.e. the final network would be robust to different spatial sizes.
